src/speed.c

```diff
@@ -194,72 +122,128 @@ void set_max_end_linear_speed(float value)
 }
 
 /**
- * @brief Turn at a defined speed level.
+ * @brief In-place turn at a defined speed level.
  *
- * @param[in] turn Turn type.
- * @param[in] speed Configured speed level.
+ * @param[in] radians Radians to turn (positive means left).
+ * @param[in] force Maximum force to apply while turning.
  */
-void speed_turn(enum movement turn, uint8_t speed)
+void inplace_turn(float radians, float force)
 {
-	parametric_turn(turns[turn][speed].max_angular_speed,
-			turns[turn][speed].rise_time,
-			turns[turn][speed].elapsed_time);
+	int32_t start;
+	int32_t current;
+	float time;
+	float angular_velocity;
+	float max_angular_velocity;
+	float factor;
+	float arc;
+	float transition;
+	float duration;
+	float transition_angle;
+
+	angular_acceleration =
+	    force * MOUSE_WHEELS_SEPARATION / MOUSE_MOMENT_OF_INERTIA;
+	max_angular_velocity = sqrt(radians / 2 * angular_acceleration);
+	if (max_angular_velocity > MOUSE_MAX_ANGULAR_VELOCITY)
+		max_angular_velocity = MOUSE_MAX_ANGULAR_VELOCITY;
+
+	duration = max_angular_velocity / angular_acceleration * PI;
+	transition_angle = duration * max_angular_velocity / PI;
+	arc = (radians - 2 * transition_angle) / max_angular_velocity;
+	transition = duration / 2;
+
+	disable_walls_control();
+	start = get_clock_ticks();
+	while (true) {
+		current = get_clock_ticks();
+		time = (float)(current - start) / SYSTICK_FREQUENCY_HZ;
+		if (time >= 2 * transition + arc)
+			break;
+		angular_velocity = max_angular_velocity;
+		if (time < transition) {
+			factor = time / transition;
+			angular_velocity *= sin(factor * PI / 2);
+		} else if (time >= transition + arc) {
+			factor = (time - arc) / transition;
+			angular_velocity *= sin(factor * PI / 2);
+		}
+		set_ideal_angular_speed(angular_velocity);
+	}
+	set_ideal_angular_speed(0);
 }
 
 /**
  * @brief Get the straight distance that a turn adds before a straight movement.
  *
- * @param[in] turn Turn type.
- * @param[in] speed Configured speed level.
+ * @param[in] turn_type Turn type.
  *
  * @return The added distance.
  */
-float get_move_turn_before(enum movement move, uint8_t speed)
+float get_move_turn_before(enum movement turn_type)
 {
-	return turns[move][speed].before;
+	return turns[turn_type].before;
 }
 
 /**
  * @brief Get the straight distance that a turn adds after a straight movement.
  *
- * @param[in] turn Turn type.
- * @param[in] speed Configured speed level.
+ * @param[in] turn_type Turn type.
  *
  * @return The added distance.
  */
-float get_move_turn_after(enum movement move, uint8_t speed)
+float get_move_turn_after(enum movement turn_type)
 {
-	return turns[move][speed].after;
+	return turns[turn_type].after;
 }
 
 /**
  * @brief Get the expected linear speed at which to turn.
  *
- * @param[in] turn Turn type.
- * @param[in] speed Configured speed level.
+ * @param[in] turn_type Turn type.
+ * @param[in] force Maximum force to apply while turning.
  *
- * @return The expected speed.
+ * @return The added distance.
  */
```

Comment:
@return the speed

Reply:
Could you explain the formula?

Reply:
https://github.com/Bulebots/bulebule/issues/388

Reply:
280c6897af00636b5eda2c38c6950c0b3d9b7230